### PR TITLE
Add SemVer2 support to Lucene Jobs

### PR DESCRIPTION
**Pull request by ryuyu**
This PR adds support for writing a new Field into the Lucene Index, SemVerLevel.
Currently, this field will match the value in the SemVerLevelKey column in the database added with https://github.com/NuGet/NuGetGallery/pull/3678

https://github.com/NuGet/NuGet.Services.Metadata/pull/142 contains parallel changes for search service to consume these updates.

@joelverhagen @skofman1 @xavierdecoster

### CODE REVIEW

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 1]
```diff
@@ -124,6 +125,42 @@ private void AddListed()
...
+                        }
+                    }
+                }
+
+                var dependencyGroups = _reader.GetPackageDependencies().ToList();
+                foreach (var dependencyGroup in dependencyGroups)
+                {
+                    if (dependencyGroup.Packages.Any())
+                    {
+                        // Add packages list
```
Comment (xavierdecoster) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: Not sure what you mean by "adding" packages list. Think the comment can be removed.

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 2]
```diff
@@ -124,6 +125,42 @@ private void AddListed()
...
+                        foreach (var packageDependency in dependencyGroup.Packages)
+                        {
+                            var versionRange = packageDependency.VersionRange;
+                            if ((versionRange.MaxVersion != null && versionRange.MaxVersion.IsSemVer2)
+                                || (versionRange.MinVersion != null && versionRange.MinVersion.IsSemVer2))
+                            {
+                                _metadata["semVerLevelKey"] = "2";
```
Comment (xavierdecoster) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: The string `"semVerLevelKey"` is used in multiple places in this PR: extract const. Same for the value `"2"`.

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 3]
```diff
@@ -124,6 +125,42 @@ private void AddListed()
...
+                    {
+                        if (packageOriginalVersion.IsSemVer2)
+                        {
+                            _metadata["semVerLevelKey"] = "2";
+                            return;
+                        }
+                    }
+                }
+
+                var dependencyGroups = _reader.GetPackageDependencies().ToList();
```
Comment (skofman1) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: consider extracting the dependency pass to a separate method for readability

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs (hunk 1 above)
Comment (ryuyu) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: Cleaned up.

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 4]
```diff
@@ -121,7 +122,42 @@ private void AddListed()
...
+                            _metadata[MetadataConstants.SemVerLevelKeyPropertyName] = MetadataConstants.SemVerLevel2Value;
+                            return;
+                        }
+                    }
+                }
+
+                var dependencyGroups = _reader.GetPackageDependencies().ToList();
+                foreach (var dependencyGroup in dependencyGroups)
+                {
+                    if (dependencyGroup.Packages.Any())
```
Comment (xavierdecoster) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: Don't think we need this additional check here, as the enumeration on [line 149](https://github.com/NuGet/NuGet.Services.Metadata/pull/144/files#diff-8e9238f17f25f40d2a5a0ef47ca1901eR149) will just yield no results.

File: src/NuGet.Indexing/Extraction/MetadataConstants.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGet.Indexing
+{
+    public static class MetadataConstants
```
Comment (xavierdecoster) [inline src/NuGet.Indexing/Extraction/MetadataConstants.cs]: 👍

Comment (xavierdecoster) [inline src/NuGet.Indexing/Extraction/MetadataConstants.cs]: Are these constants used outside this `NuGet.Indexing` library? If not, make it `internal`.

File: tests/NuGet.IndexingTests/Extraction/CatalogPackageMetadataExtractorTests.cs [hunk 6]
```diff
@@ -23,6 +23,24 @@ public void AddsListed(object catalogEntry, string expected)
...
+        {
+            // Arrange
+            var catalogEntryJObject = CatalogEntry(catalogEntry);
+
+            // Act
+            var metadata = CatalogPackageMetadataExtraction.MakePackageMetadata(catalogEntryJObject);
+
+
+            // Assert
+            Assert.Equal(expectedToContainKey, metadata.Keys.Contains("semVerLevelKey"));
```
Comment (xavierdecoster) [inline tests/NuGet.IndexingTests/Extraction/CatalogPackageMetadataExtractorTests.cs]: Use the `LuceneMetadata.SemVerLevelKeyPropertyName` constant here (and throughout these tests).

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs (hunk 4 above)
Comment (ryuyu) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: Done

File: tests/NuGet.IndexingTests/Extraction/CatalogPackageMetadataExtractorTests.cs (hunk 6 above)
Comment (ryuyu) [inline tests/NuGet.IndexingTests/Extraction/CatalogPackageMetadataExtractorTests.cs]: done

**Pull request merged**